helix-lsp/src/snippet.rs

```diff
@@ -86,17 +83,34 @@ mod parser {
         else        ::= text
     */
 
-    static DIGIT: Lazy<regex::Regex> = Lazy::new(|| regex::Regex::new(r"^[0-9]+").unwrap());
-    static VARIABLE: Lazy<regex::Regex> =
-        Lazy::new(|| regex::Regex::new(r"^[_a-zA-Z][_a-zA-Z0-9]*").unwrap());
-    static TEXT: Lazy<regex::Regex> = Lazy::new(|| regex::Regex::new(r"^[^\$]+").unwrap());
-
     fn var<'a>() -> impl Parser<'a, Output = &'a str> {
-        pattern(&VARIABLE)
+        // var = [_a-zA-Z][_a-zA-Z0-9]*
+        move |input: &'a str| match input
+            .char_indices()
+            .take_while(|(p, c)| {
+                *c == '_'
+                    || if *p == 0 {
+                        c.is_ascii_alphabetic()
+                    } else {
+                        c.is_ascii_alphanumeric()
+                    }
+            })
+            .last()
+        {
+            Some((index, c)) if index >= 1 => {
+                let index = index + c.len_utf8();
+                Ok((&input[index..], &input[0..index]))
+            }
+            _ => Err(input),
+        }
+    }
+
+    fn txt<'a>() -> impl Parser<'a, Output = &'a str> {
+        take_while(|c| c != '$')
     }
 
     fn digit<'a>() -> impl Parser<'a, Output = usize> {
-        filter_map(pattern(&DIGIT), |s| s.parse().ok())
+        filter_map(take_while(|c| c.is_ascii_digit()), |s| s.parse().ok())
     }
```

Comment:
Nice work on dropping the regex dependency here! Might worth benchmarking whether it actually performs better though, I know that the `regex` crate will compile simple regex state machines to something that will be similar in performance to a well optimized manual implementation.

Unlikely to make an actual real world difference though

Reply:
Yeah I definitely prefer this to regex. In my experience there is no practical performance difference between `regex` and a handwritten lexer. Compilers (as in `rustc`/`llvm`) are **really** good at their job so often the handwritten version actually performs better because `llvm` can perform efficient register allocation (although sometimes abstractions like iterator can cause overhead if `llvm` fails to optimized them out but it's usually not hard to optimize handwritten code to match regex).

In anycase the difference likely wouldn't matter unless as server sends 100MB snippets.

Reply:
As regex cannot (yet) be compiled at compile time many optimizations cannot be done at compile time and thus an "handwritten" parse will almost always be faster. The difference in performance is probably not that big given that snippets are generally not that long (<50 characters) but those regex are called many times and since regex have an "startup time" that is mostly constant being able to remove it leads almost certainly to better latency, which is particularly important when selecting a completion snippet or when switching very fast between completions.

There is also the fact that having one less dependency on regex is a good thing.